### PR TITLE
AC-477 Update VisitDAO for Room

### PR DESCRIPTION
**Pull request by VibhorChinda**
## Description of what I changed
Created VisitRoomDAO Interface for implementing Room methods.

## Issue I worked on
JIRA Issue: https://issues.openmrs.org/browse/AC-477

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [X] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by VibhorChinda** [comment]
@f4ww4z  please have a look at the PR

**Comment by VibhorChinda** [comment]
> @VibhorChinda Use [VisitEntity](https://github.com/openmrs/openmrs-contrib-android-client/blob/master/openmrs-client/src/main/java/org/openmrs/mobile/databases/entities/VisitsEntity.java) instead of `Visit` class, we'll be removing it once finished migrating.

@f4ww4z sir please have a look at the changed queries and can you suggest any article for writing tests. I was trying to write the tests from the article you linked in my other PR, but faced problem while getting the context in that test file. Any suggestions sir ??

**Comment by f4ww4z** [comment]
@VibhorChinda Can you write the failing tests you mentioned? Then I can take a look.

**Comment by VibhorChinda** [comment]
@f4ww4z sir please have a look. I am facing error on getting context in my unit text.

File: openmrs-client/src/main/res/layout/fragment_login.xml [hunk 1]
```diff
@@ -58,7 +58,7 @@
                 android:layout_marginBottom="10dp"
                 android:gravity="center"
                 android:orientation="vertical"
-                android:visibility="gone">
+                android:visibility="visible">
```
Comment (f4ww4z) [inline openmrs-client/src/main/res/layout/fragment_login.xml]: How is this change related to the PR?

File: openmrs-client/src/test/java/org/openmrs/mobile/test/VisitRoomDAOTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.openmrs.mobile.databases.AppDatabase;
+import org.openmrs.mobile.databases.entities.VisitsEntity;
+
+import rx.Scheduler;
+import rx.Single;
+import rx.SingleSubscriber;
+import rx.Subscriber;
```
Comment (f4ww4z) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/VisitRoomDAOTest.java]: These 4 imports are unused.

**Comment by f4ww4z** [comment]
> @VibhorChinda You need to [see this post](https://stackoverflow.com/a/36445307/8434188). You cannot get the context in a local unit test, you would get it in an instrumented test. Hence you need to write instrumented tests (in `androidTest` folder), not unit tests.

The same is true for #636

File: openmrs-client/src/main/java/org/openmrs/mobile/databases/entities/VisitsEntity.java [hunk 3]
```diff
@@ -42,7 +42,7 @@ public void setVisitPlace(String visitPlace) {
         this.visitPlace = visitPlace;
     }
 
-    public void setStartDate(boolean startDate) {
+    public void setStartDate(String startDate) {
```
Comment (ribhavsharma) [inline openmrs-client/src/main/java/org/openmrs/mobile/databases/entities/VisitsEntity.java]: Can you add a getter for startDate too? I see it's one of the reasons why the Travis CI is failing.

File: openmrs-client/src/main/java/org/openmrs/mobile/dao/VisitRoomDAO.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    Single<VisitsEntity> getVisitByID(final Long visitID);
+
+    @Query("SELECT * FROM visits WHERE uuid = :visitUUID")
+    Single<Long> getVisitsIDByUUID(final String visitUUID);
+
+    @Query("SELECT * FROM visits WHERE uuid = :uuid")
+    Single<VisitsEntity> getVisitByUuid(String uuid);
+
+    @Delete
+    Observable<Boolean> deleteVisitsByPatientId(Long id);
```
Comment (ribhavsharma) [inline openmrs-client/src/main/java/org/openmrs/mobile/dao/VisitRoomDAO.java]: I see that the Travis CI also shows an error with this line.
`error: Type of the parameter must be a class annotated with @Entity or a collection/array of it.`
I believe this could be fixed by changing the class to an Abstract Class.

File: openmrs-client/src/main/res/layout/fragment_login.xml (hunk 1 above)
Comment (VibhorChinda) [reply openmrs-client/src/main/res/layout/fragment_login.xml]: will remove this change from here @f4ww4z sir.

**Comment by VibhorChinda** [comment]
@f4ww4z sir please have a look at the PR.
again the tests are failing.

File: openmrs-client/src/androidTest/java/org/openmrs/mobile/test/VisitRoomDAOTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        });
+
+    }
+
+    @After
+    public void closeDb() throws Exception {
+        database.close();
+    }
+
+}
```
Comment (Devanshk9) [inline openmrs-client/src/androidTest/java/org/openmrs/mobile/test/VisitRoomDAOTest.java]: @VibhorChinda sir please add a new line at the end of files.
read more at- https://www.quora.com/Why-is-it-a-common-practice-to-put-a-new-line-at-the-end-of-a-source-code-file

Comment (PrathameshMutkure) [reply openmrs-client/src/androidTest/java/org/openmrs/mobile/test/VisitRoomDAOTest.java]: Hii @Devanshk9, The new line is already present there, GitHub doesn't render the last new line but it does point out missing new line at end with a red symbol, which in this case is absent.

So, no problem in here

**Comment by VibhorChinda** [comment]
@f4ww4z sir can you please have a look at the PR. Unit test is working now too.

**Comment by VibhorChinda** [comment]
Closing this pull request. As i have made some changes which were unrelated to this issue.
Here is the link to new PR https://github.com/openmrs/openmrs-contrib-android-client/pull/670

**Comment by VibhorChinda** [comment]
Closing this pull request. As i have made some changes which were unrelated to this issue.
Here is the link to new PR #670
